File: android/app/src/main/python/stock_monitoring_app/indicators/ma_indicator.py
import pandas as pd
from .base_indicator import Indicator

class MAIndicator(Indicator):
    """
    Calculates Simple Moving Average (SMA) or Exponential Moving Average (EMA) using pandas.
    """


    @staticmethod
    def get_search_space():
        return {
            # Try shorter windows more suitable for intraday 1-minute data (288 points)
            "window": [5, 8, 13, 21, 34], 
            "ma_type": ["sma", "ema"],
            "column": ["Close"] # Focusing on 'Close' for simplicity initially
        }

    def __init__(
        self,
        df: pd.DataFrame,
        window: int = 20,
        ma_type: str = "sma",
        column: str = "Close"
    ):
        """
        Args:
            df: OHLCV DataFrame.
            window: The time period for the moving average.
            ma_type: 'sma' for Simple, 'ema' for Exponential.
            column: Which column to calculate MA on.
        """

        super().__init__(df)
        self.window = window

        # Check if DataFrame has enough data for the given window.
        if len(self.df) < self.window:
            raise ValueError(
                f"Insufficient data for {self.__class__.__name__} (window: {self.window}): "
                f"{len(self.df)} rows provided, requires at least {self.window} rows."
            )
        self.ma_type = ma_type.lower()
        self.column = column

        # Signal names for MA crosses
        self.ma_col_name = f"{self.ma_type.upper()}_{self.window}_{self.column}"
        self.buy_signal_col = f"{self.ma_col_name}_Cross_Above"
        self.sell_signal_col = f"{self.ma_col_name}_Cross_Below"

        self.signal_orientations[self.buy_signal_col] = 'buy'
        self.signal_orientations[self.sell_signal_col] = 'sell'


    def calculate(self) -> pd.DataFrame:
        """
        Calculates MA and adds 'SMA_<window>_<column>' or 'EMA_<window>_<column>' column.
        Also generates buy/sell signals when the price crosses the MA.
        """
        price_source = self.df[self.column] # Price source for MA calculation and crosses

        if self.ma_type == "sma":
            ma_series = price_source.rolling(window=self.window, min_periods=self.window).mean()
        elif self.ma_type == "ema":
            ma_series = price_source.ewm(span=self.window, min_periods=self.window, adjust=False).mean()
        else:
            raise ValueError(f"Unknown ma_type: {self.ma_type}")

        self.df[self.ma_col_name] = ma_series

        # Generate cross signals
        # Price crosses above MA (buy signal)
        self.df[self.buy_signal_col] = (price_source > ma_series) & (price_source.shift(1) <= ma_series.shift(1))
        # Price crosses below MA (sell signal)
        self.df[self.sell_signal_col] = (price_source < ma_series) & (price_source.shift(1) >= ma_series.shift(1))

        # Ensure boolean type and fill NaNs (especially at the beginning due to shift)
        self.df[self.buy_signal_col] = self.df[self.buy_signal_col].fillna(False).astype(bool)
        self.df[self.sell_signal_col] = self.df[self.sell_signal_col].fillna(False).astype(bool)

        return self.df

File: src/app/ma_indicator.cpp
#include "app/ma_indicator.hpp"
#include "app/indicator_factory.hpp"
#include <algorithm>
#include <cmath>
#include <stdexcept>

MAIndicator::MAIndicator(const DataFrame& df, int window, const std::string& ma_type, const std::string& column)
    : Indicator(df), window_(window), ma_type_(ma_type), column_(column)
{
    size_t n = 0;
    if (column_ == "Close") n = df_.close.size();
    else if (column_ == "Open") n = df_.open.size();
    else if (column_ == "High") n = df_.high.size();
    else if (column_ == "Low") n = df_.low.size();
    else if (column_ == "Volume") n = df_.volume.size();
    else throw std::invalid_argument("Unknown column for MA calculation: " + column_);

    if (n < static_cast<size_t>(window_)) {
        throw std::invalid_argument("Insufficient data for MAIndicator (window: " +
            std::to_string(window_) + "): " + std::to_string(n) +
            " rows provided, requires at least " + std::to_string(window_) + " rows.");
    }

    std::string ma_type_uc = ma_type_;
    std::transform(ma_type_uc.begin(), ma_type_uc.end(), ma_type_uc.begin(), ::toupper);
    ma_col_name_ = ma_type_uc + "_" + std::to_string(window_) + "_" + column_;
    buy_signal_col_ = ma_col_name_ + "_Cross_Above";
    sell_signal_col_ = ma_col_name_ + "_Cross_Below";

    signal_orientations_[buy_signal_col_] = "buy";
    signal_orientations_[sell_signal_col_] = "sell";
}

std::vector<double> MAIndicator::calculate_sma(const std::vector<double>& data, int window) {
    std::vector<double> result(data.size(), std::nan(""));
    if (window <= 0 || data.size() < static_cast<size_t>(window)) return result;
    double sum = 0.0;
    for (size_t i = 0; i < data.size(); ++i) {
        sum += data[i];
        if (i >= static_cast<size_t>(window)) sum -= data[i - window];
        if (i + 1 >= static_cast<size_t>(window))
            result[i] = sum / window;
    }
    return result;
}

std::vector<double> MAIndicator::calculate_ema(const std::vector<double>& data, int window) {
    std::vector<double> result(data.size(), std::nan(""));
    if (window <= 0 || data.size() < static_cast<size_t>(window)) return result;
    double alpha = 2.0 / (window + 1.0);
    double ema = 0.0;
    for (size_t i = 0; i < data.size(); ++i) {
        if (i == window - 1) {
            double sum = 0.0;
            for (int j = 0; j < window; ++j) sum += data[j];
            ema = sum / window;
            result[i] = ema;
        } else if (i >= static_cast<size_t>(window)) {
            ema = alpha * data[i] + (1.0 - alpha) * ema;
            result[i] = ema;
        }
    }
    return result;
}

DataFrame MAIndicator::calculate() {
    const std::vector<double>* price_ptr = nullptr;
    if (column_ == "Close") price_ptr = &df_.close;
    else if (column_ == "Open") price_ptr = &df_.open;
    else if (column_ == "High") price_ptr = &df_.high;
    else if (column_ == "Low") price_ptr = &df_.low;
    else if (column_ == "Volume") price_ptr = &df_.volume;
    else throw std::invalid_argument("Unknown column for MA calculation: " + column_);
    const std::vector<double>& price = *price_ptr;
    size_t n = price.size();

    std::vector<double> ma_series;
    if (ma_type_ == "sma") {
        ma_series = calculate_sma(price, window_);
    } else if (ma_type_ == "ema") {
        ma_series = calculate_ema(price, window_);
    } else {
        throw std::invalid_argument("Unknown ma_type: " + ma_type_);
    }

    std::vector<bool> buy_signal(n, false);
    std::vector<bool> sell_signal(n, false);

    for (size_t i = 1; i < n; ++i) {
        if (!std::isnan(ma_series[i]) && !std::isnan(ma_series[i-1])) {
            if (price[i] > ma_series[i] && price[i-1] <= ma_series[i-1])
                buy_signal[i] = true;
            if (price[i] < ma_series[i] && price[i-1] >= ma_series[i-1])
                sell_signal[i] = true;
        }
    }

    DataFrame out = df_;
    out.extra_columns[ma_col_name_] = ma_series;
    out.extra_bool_columns[buy_signal_col_] = buy_signal;
    out.extra_bool_columns[sell_signal_col_] = sell_signal;
    return out;
}

REGISTER_INDICATOR("MA",MAIndicator)
